Implement Fund Card component

rustam-cb opened the pull request:
**What changed? Why?**
Implemented a new component `FundCard`
P/PS: https://docs.google.com/document/d/1y57fxsga2ZLF8-ieH2OY7jvqx5468txQuek1WTYPBxE/edit?tab=t.0

Usage: 
```
FundCard assetSymbol="ETH" /> 
```

Component is customizable and there are few components that can be passed into it.

```
  AmountInputComponent 
  HeaderComponent
  AmountInputTypeSwithComponent
  PaymentMethodSelectorDropdownComponent
  SubmitButtonComponent
```

|Fiat input|Crypto input|
|----------|------------|
|<img width="455" alt="Screenshot 2024-12-17 at 9 38 52 PM" src="https://github.com/user-attachments/assets/52161ad1-0b59-4dd8-8019-c9578c3fa5d7" />|<img width="451" alt="Screenshot 2024-12-17 at 9 39 06 PM" src="https://github.com/user-attachments/assets/8c8e5097-58f6-4cf9-9cdb-268575d9ce99" />|


|No input (default) | payment method dropdown|
|-------------------|--------------------------|
|<img width="452" alt="Screenshot 2024-12-17 at 9 38 41 PM" src="https://github.com/user-attachments/assets/0a24e1d9-8924-429a-bcde-f40a6b9c5861" />|<img width="451" alt="Screenshot 2024-12-17 at 9 39 20 PM" src="https://github.com/user-attachments/assets/03b92ad5-5b0a-40ea-9519-e8d12d5c7435" />|





**Notes to reviewers**
Overall component structure:
- FundCard
- - HeaderComponent
- - AmountInputComponent
- - PaymentMethodDropdownComponent
- - SubmitButtonComponent


**How has it been tested?**
- Unit tests
- Manual testing

dschlabach commented on playground/nextjs-app-router/components/demo/Swap.tsx:
```diff
@@ -45,9 +45,9 @@ function SwapComponent() {
 
   const usdcToken: Token = {
-    name: 'USDC',
+    name: 'USD',
     address: '0x833589fcd6edb6e08f4c7c32d4f71b54bda02913',
-    symbol: 'USDC',
+    symbol: 'USD',
     decimals: 6,
     image:
```
Should revert this or else our Swap component won't show USDC as needed:

![Uploading CleanShot 2024-12-18 at 12.57.24@2x.png…]()

rustam-cb replied:
Been testing it and forgot to revert. Thank you for catching!

dschlabach commented on vitest.config.ts:
```diff
@@ -24,6 +24,7 @@ export default defineConfig({
         'playground/**',
         'site/**',
         'create-onchain/**',
+        '**/**.test.tsx',
```
Intentional?

rustam-cb replied:
Yes. The code coverage was listing test files and counting them towards coverage %.  I had couple of my test files noted as covered 99.97% which prevented me from creating PR. (Its likely due to helper objects, functions that you create in a test files).

dschlabach commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,34 +21,65 @@ export function FundButton({
     componentTheme,
-    pressable.primary,
+    buttonColorClass,
     'px-4 py-3 inline-flex items-center justify-center space-x-2 disabled',
```
Probably want to add hover:cursor-disabled here

dschlabach replied:
Sorry, meant cursor-not-allowed https://tailwindcss.com/docs/cursor#setting-the-cursor-style

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,197 @@
+            componentTheme,
+            text.body,
+            'border-[none] bg-transparent',
+            'text-[60px] leading-none outline-none',
```
Can we make this rem based? Otherwise won't scale properly

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,197 @@
+          />
+        )}
+
+        <input
```
It would be nice if the input spanned the entire width, otherwise if a user clicks to the right, they won't focus the input
![CleanShot 2024-12-18 at 13 31 46@2x](https://github.com/user-attachments/assets/7d33467a-57b3-470d-9be7-7be4b1685f74)

rustam-cb replied:
Unfortunately currently it is working this way because if you switch to Crypto it will have a space [1 - - - - - - -ETH] but we want the asset name next to the digits [1 ETH].  
But you are right, we should make it better for users. I'll take a look into focusing on the input field when that area is clicked. (I still have a few features to add to this component, so I'll make this change in a follow up PR )

dschlabach commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,95 @@
+import { FundCardPaymentMethodSelectorToggle } from './FundCardPaymentMethodSelectorToggle';
+import { useFundContext } from './FundCardProvider';
+
+export function FundCardPaymentMethodDropdown({
```
Should be able to close on Esc and either close when user tabs out of the dropdown (or trap focus).

Sometimes I also get this flash

https://github.com/user-attachments/assets/5e3bb4a1-f542-4cff-a079-f23cd3b34746

dschlabach commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,95 @@
+            'ock-scrollbar',
+          )}
+        >
+          <div className={cn(background.inverse, 'p-2', 'overflow-y-auto')}>
```
```suggestion
          <div className={cn(background.inverse, 'p-2 overflow-y-auto')}>
```

dschlabach commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,34 +21,65 @@ export function FundButton({
     componentTheme,
-    pressable.primary,
+    buttonColorClass,
     'px-4 py-3 inline-flex items-center justify-center space-x-2 disabled',
```
is `disabled` a class in our codebase? I can't find it

rustam-cb replied:
not sure. I did not add it but I'll remove it. Does not seem to have any effect/

cpcramer commented:
Seeing a strange flicker when using the dropdown

![Clipboard-20241218-192142-422](https://github.com/user-attachments/assets/bb811b04-78e0-4756-a93f-fbedc7e8d883)

cpcramer commented on src/core-react/internal/hooks/useIcon.tsx:
```diff
@@ -1,26 +1,38 @@
     switch (icon) {
       case 'coinbasePay':
-        return coinbasePaySvg;
+        return <CoinbasePaySvg className={className} />;
```
nit: can we return the svg similar to the other cases? 
```suggestion
        return coinbasePaySvg;
```

rustam-cb replied:
Is it possible to use className on svg? The reason is that I need to pass a color to the svg.

dschlabach replied:
Btw you can use `fill="currentColor"` on the SVG and then set `text-blue-400` or whatever you want with Tailwind

dschlabach replied:
I would like to have this standard so that when you call `useIcon` it behaves the same regardless of any icon that you pass in. 

I understand you want to change the color of this CoinbasePay icon, but we should have one standard interface for all the SVGs or else there are going to be unintentional cases of icons breaking

cpcramer commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,154 @@
+import { FundCardProvider } from './FundCardProvider';
+import { useFundContext } from './FundCardProvider';
+
+const defaultPaymentMethods: PaymentMethodReact[] = [
```
```suggestion
const DEFAULT_PAYMENT_METHODS: PaymentMethodReact[] = [
```

cpcramer replied:
Can we add this to a constants file?

cpcramer commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,154 @@
+  );
+}
+
+export function FundCardContent({
```
Does this need to be exported?

rustam-cb replied:
nope, it does not.

cpcramer commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Ensure the decimal value is formatted correctly (i.e. "0.1" instead of ".1" and "0.1" instead of "01")
+ */
+const formatDecimalInputValue = (value: string) => {
```
This component is getting pretty complex - I feel like we should try to export some functionality into utility functions.

cpcramer commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+  exchangeRate,
+  isLoading,
+}: FundCardAmountInputTypeSwitchPropsReact) => {
+  const componentTheme = useTheme();
```
`componentTheme` should only have to be applied once to the top level component

cpcramer commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+    setSelectedInputType(selectedInputType === 'fiat' ? 'crypto' : 'fiat');
+  };
+
+  const formatUSD = useCallback((amount: string) => {
+    if (!amount || amount === '0') {
+      return null;
+    }
+    const roundedAmount = Number(getRoundedAmount(amount, 2));
+    return `$${roundedAmount.toFixed(2)}`;
+  }, []);
```
I've seen this function a few times, we should a reusable one cc @dschlabach @alissacrane-cb

cpcramer commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+        className={cn(
+          text.label2,
+          color.foregroundMuted,
+          'font-normal',
```
Is this necessary? I think it defaults to font-normal?

rustam-cb replied:
Removing it makes the text `bold`

cpcramer commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+        aria-label="amount type switch"
+        className={cn(
+          pressable.default,
+          'mr-1 rounded-full p-1 opacity-50 transition-opacity hover:opacity-100',
```
Can use the `border.radius` class here

cpcramer commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+      >
+        <div className="h-[1.125rem] w-[1.125rem]">{iconSvg}</div>
+      </button>
+      <div style={textStyle}>
+        {selectedInputType === 'fiat' ? cryptoAmountLine : fiatAmountLine}
+
+        {exchangeRateLine}
+      </div>
+    </div>
+  );
+};
+
+const textStyle = {
+  textOverflow: 'ellipsis',
+  width: '390px',
+  overflow: 'hidden',
+  whiteSpace: 'nowrap',
+};
```
Let's try to avoid inline styling in favor of Tailwind class based styling,

cpcramer commented on src/fund/components/FundCardCurrencyLabel.tsx:
```diff
@@ -0,0 +1,26 @@
+  HTMLSpanElement,
+  FundCardCurrencyLabelPropsReact
+>(({ currencySign }, ref) => {
+  const componentTheme = useTheme();
```
Same as above but you should only need to pass this componentTheme down in the top level component

cpcramer commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,95 @@
+    setSelectedPaymentMethod(paymentMethods[0]);
+  }, []);
+
+  useEffect(() => {
+    // Add event listener for outside clicks
+    document.addEventListener('click', handleBlur);
+    return () => {
+      // Clean up the event listener
+      document.removeEventListener('click', handleBlur);
+    };
+  }, [handleBlur]);
```
We have `useOutsideClick` hook that should help with this

cpcramer commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,95 @@
+            componentTheme,
+            border.radius,
+            'absolute right-0 z-10 mt-1 flex max-h-80 w-full flex-col overflow-y-hidden',
+            'ock-scrollbar',
```
What's this?

cpcramer commented on src/fund/components/FundCardPaymentMethodImage.tsx:
```diff
@@ -0,0 +1,59 @@
+        data-testid="fundCardPaymentMethodImage__noImage"
+        style={styles.image}
+      >
+        <div style={styles.placeholderImage} />
```
I'd prefer to avoid inline styling but curious what the team thinks

cpcramer commented on src/fund/components/FundCardPaymentMethodSelectRow.tsx:
```diff
@@ -0,0 +1,48 @@
+        className={cn(
+          componentTheme,
+          pressable.default,
+          'ock-border-radius flex w-full items-center justify-between px-2 py-1',
```
Use `border.radius` from theme.ts instead

cpcramer commented on src/internal/components/Skeleton.tsx:
```diff
@@ -0,0 +1,19 @@
+  className?: string;
+};
+
+export function Skeleton({ className }: SkeletonReact) {
```
What is this `Skeleton` component for?

cpcramer commented on src/internal/components/Skeleton.tsx:
```diff
@@ -0,0 +1,19 @@
+      className={cn(
+        'animate-pulse bg-opacity-50',
+        background.alternate,
+        'rounded',
```
Would` border.*` work from theme.ts?

alessey commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,47 +21,114 @@ export function FundButton({
+          return null;
+        }
+
+        const interval = setInterval(() => {
```
possible to use this?  https://github.com/coinbase/onchainkit/blob/main/src/buy/hooks/usePopupMonitor.ts

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,134 @@
+          submitButtonComponent={submitButtonComponent}
+        />
+      </div>
+    </FundCardProvider>
```
the rest of our components are set up to accept child components vs props

alessey replied:
any reason not to do the same here?

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+        const resultCryptoValue = truncateDecimalPlaces(
+          calculatedCryptoValue,
+          8,
+        );
```
nit: you could just use toFixed on your calculatedCryptoValue before casting it to a string, below as well for fiat

brendan-defi replied:
i use this in the new Wallet component we are releasing this week, can you use this?

```
new Intl.NumberFormat('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 5,
  })
```

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+            -moz-appearance: textfield;
+          }
+        `}
+      </style>
```
could you move this to a tailwind utility instead of an inline style

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+      // Set the input width based on the span width
+      if (inputRef.current) {
+        inputRef.current.style.width = `${width}px`;
+        inputRef.current.style.maxWidth = `${390 - currencyWidth}px`;
```
what are you trying to do here?    Something like this may be more useful: https://github.com/coinbase/onchainkit/pull/1224

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+
+          With this span we can measure the width of the text in the input field and set the width of the input field to match the text width
+          [0.12][ETH] - Now the currency symbol is displayed next to the input field
+      */}
```
can't the input just scale as needed with a max-width?

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+        ref={hiddenSpanRef}
+        className={cn(
+          text.body,
+          'border-[none] bg-transparent',
```
border-none

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,167 @@
+        <input
+          className={cn(
+            text.body,
+            'border-[none] bg-transparent',
```
border-none

alessey commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,84 @@
+
+  const fiatAmountLine = useMemo(() => {
+    return <span className={cn(text.label1)}>{formatUSD(fundAmountFiat)}</span>;
+  }, [formatUSD, fundAmountFiat]);
```
nit: this would be better as a single useMemo that returned the correctAmount line based on type

alessey commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,92 @@
+        ref={buttonRef}
+        onClick={handleToggle}
+        isOpen={isOpen}
+        paymentMethod={selectedPaymentMethod || paymentMethods[0]}
```
why is this needed if you are doing this in the useEffect, or can you remove the useEffect and do this instead?

alessey commented on src/fund/components/FundCardPaymentMethodImage.tsx:
```diff
@@ -0,0 +1,59 @@
+        'flex items-center justify-center overflow-hidden rounded-[50%]',
+        className,
+      )}
+      style={styles.image}
```
use tailwind

alessey commented on src/fund/components/FundCardPaymentMethodImage.tsx:
```diff
@@ -0,0 +1,59 @@
+        height: `${size}px`,
+        minWidth: `${size}px`,
+        minHeight: `${size}px`,
+      },
```
anything defined in px won't scale properly with browser zoom levels

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,171 @@
+        `}
+      </style>
+
+      <div className="flex" style={{ height: '78px' }}>
```
Stick to tailwind classes for consistency

dschlabach replied:
preferably `h-20` here

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,171 @@
+          className={cn(
+            text.body,
+            'border-none bg-transparent',
+            'text-[3.75rem] leading-none outline-none',
```
should be `text-6xl` instead of arbitrary

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,171 @@
+        className={cn(
+          text.body,
+          'border-none bg-transparent',
+          'text-[3.75rem] leading-none outline-none',
```
`text-6xl`

dschlabach commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,171 @@
+          'border-none bg-transparent',
+          'text-[3.75rem] leading-none outline-none',
+        )}
+        style={{
+          position: 'absolute',
+          opacity: 0,
+          whiteSpace: 'nowrap',
+          pointerEvents: 'none',
+        }}
```
move into Tailwind utilities for consistency

rustam-cb replied:
@dschlabach  Where is the Tailwing utilities? I can't find any in the codebase. (The css I see is autogenerated and I can't add anything there)

cpcramer replied:
All Tailwind utilities are accessible anywhere in /src. You can apply them directly as classes. https://tailwindcss.com/docs/utility-first

dschlabach commented on src/internal/svg/successSvg.tsx:
```diff
@@ -11,7 +13,7 @@ export const successSvg = (
     <path
       d="M8 0C3.58 0 0 3.58 0 8C0 12.42 3.58 16 8 16C12.42 16 16 12.42 16 8C16 3.58 12.42 0 8 0ZM6.72667 11.5333L3.73333 8.54L4.67333 7.6L6.72667 9.65333L11.44 4.94L12.38 5.88L6.72667 11.5333Z"
-      fill="#65A30D"
+      className={className}
```
This will break color styling

rustam-cb replied:
@dschlabach  It used to be hardcoded `65A30D`, and now I'm using `icon.success`. Do you mean that `icon.success` is a different tone than `65A30D` ?

cpcramer replied:
This is likely hardcoded because the icon fill color should not update based off of themes / modes

dschlabach commented on src/internal/svg/errorSvg.tsx:
```diff
@@ -11,7 +13,7 @@ export const errorSvg = (
     <path
       d="M8 16C12.4183 16 16 12.4183 16 8C16 3.58171 12.4183 0 8 0C3.58172 0 0 3.58171 0 8C0 12.4183 3.58172 16 8 16ZM11.7576 5.0909L8.84853 8L11.7576 10.9091L10.9091 11.7576L8 8.84851L5.09093 11.7576L4.2424 10.9091L7.15147 8L4.2424 5.0909L5.09093 4.24239L8 7.15145L10.9091 4.24239L11.7576 5.0909Z"
-      fill="#E11D48"
+      className={className}
```
This will break color styling

dschlabach commented on src/internal/svg/coinbasePaySvg.tsx:
```diff
@@ -12,7 +14,7 @@ export const coinbasePaySvg = (
       clipRule="evenodd"
       d="M10.0145 14.1666C7.82346 14.1666 6.04878 12.302 6.04878 9.99996C6.04878 7.69788 7.82346 5.83329 10.0145 5.83329C11.9776 5.83329 13.6069 7.33677 13.9208 9.30552H17.9163C17.5793 5.02774 14.172 1.66663 10.0145 1.66663C5.63568 1.66663 2.08301 5.39926 2.08301 9.99996C2.08301 14.6007 5.63568 18.3333 10.0145 18.3333C14.172 18.3333 17.5793 14.9722 17.9163 10.6944H13.9208C13.6069 12.6632 11.9776  ...
-      fill="#f9fafb"
+      className={className}
```
Doesn't this break default styling in things like `CheckoutButton`?

rustam-cb replied:
I have set color `icon.foreground` by default

cpcramer replied:
This will break the default styling for this SVG when themes are applied. If you need a CoinbasePaySVG to update then we might have to create a new SVG for that

dschlabach commented on src/internal/components/TextInput.tsx:
```diff
@@ -49,6 +51,7 @@ export function TextInput({
 
   return (
     <input
+      style={style}
```
Why do we need to add an inline style attribute here? We should try to stick to `className` for consistency

dschlabach commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,164 @@
+  paymentMethods = DEFAULT_PAYMENT_METHODS,
+  submitButtonComponent,
+}: FundCardContentPropsReact) {
+  /**
+   * Fetches and sets the exchange rate for the asset
+   */
+  useExchangeRate(assetSymbol);
```
Should this be moved into the provider or is there a reason it's here? A bit hard to tell what side effect it's having

cpcramer commented:
I recommend splitting this into smaller PRs, it will make reviewing and updating a lot easier and faster!

rustam-cb commented:
I have addressed all the comments. I'm not moving forward with splitting the PR up since its already had a few iterations of the review and splitting would mean each PR would need to be reviewed again. Moving forward I'll make sure to not have such big PRs.

dschlabach commented on src/fund/components/FundCardAmountInput.test.tsx:
```diff
@@ -0,0 +1,111 @@
+import { FundCardAmountInput } from './FundCardAmountInput';
+
+describe('FundCardAmountInput', () => {
+  const defaultProps = {
+    fiatValue: '100',
+    setFiatValue: vi.fn(),
+    cryptoValue: '0.05',
+    setCryptoValue: vi.fn(),
+    currencySign: '$',
+    assetSymbol: 'ETH',
+    inputType: 'fiat',
+    exchangeRate: '2',
+  } as unknown as FundCardAmountInputPropsReact;
```
Is it possible to use type annotation instead of type assertion here? That will prevent us from forgetting to update this object in case `FundCardAmountInputPropsReact` changes in the future

e.g. `const defaultProps: FundCardAmountInputPropsReact`, or does that break because of the `vi.fn()`?

dschlabach replied:
@rustam-cb - does that not work or is there a reason you resolved this comment? We just made many of our tests typesafe and would like to keep that pattern going forward

dschlabach commented on src/fund/components/FundCardHeader.tsx:
```diff
@@ -0,0 +1,18 @@
+
+  return (
+    <div
+      className={cn('font-display text-[16px]', 'leading-none outline-none')}
```
`text-[16px]` should be `text-base` and no need to call `cn()` here, can be a single string

dschlabach commented on src/fund/components/FundCardCurrencyLabel.tsx:
```diff
@@ -0,0 +1,22 @@
+      className={cn(
+        text.body,
+        'flex items-center justify-center bg-transparent',
+        'text-[60px] leading-none outline-none',
```
`text-6xl` is 60px by default, but it's in rem so scales better if the user's default font size is different 

https://tailwindcss.com/docs/font-size

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,171 @@
+        inputRef.current.style.maxWidth = `${390 - currencyWidth}px`;
+      }
+    }
+  }, [value]);
```
can you just use two inline elements here and let the browser determine the width?

rustam-cb replied:
Can't do this. 
The reason I have this is because I have 2 cases: 
First:
[$ 50            ]

Second: 
[50 ETH        ]

For the second case I need the ETH to be next to the text. All other solutions I tried makes the ETH to be at the end of the input like so: 
[50         ETH]

alessey replied:
inline elements have a width equal to their contents.  instead of using a flex box to align block level elements, just use inline and let the browser do the work, https://www.w3schools.com/html/html_blocks.asp

rustam-cb replied:
@alessey 
- I have an inline element which is hidden and I have input element. 
- Input does not resize based on its content.  
- When you enter text into the input it also gets added to the inline element and then I take the inline element width and set it to the input. 

Could you please help me understand the solution with two inline elements that you are suggesting?

alessey replied:
since there are no borders or anything, you can  have your input and "labels" on different layers, grid, etc.  you can use your hidden span to be the spacer for the currency without actually resizing the input using js, but that may be less flexible if we want to reuse since it depends on styling.  This solution is fine, ideally we'd extract to an `<AutoresizeInput/>` component.

cpcramer commented:
Buy ETH text is not updating on theme change: 
<img width="491" alt="Screenshot 2025-01-06 at 2 06 38 PM" src="https://github.com/user-attachments/assets/169ac184-2e46-4800-af39-92167dbc3af8" />

cpcramer commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +22,121 @@ export function FundButton({
+    return (
+      <>
+        {buttonIcon && (
+          // h-6 is to match the icon height to the line-height set by text.headline
```
Nit: can probably remove this comment

rustam-cb replied:
@cpcramer  Done

cpcramer commented on src/styles/theme.ts:
```diff
@@ -26,6 +26,7 @@ export const pressable = {
     'cursor-pointer ock-bg-inverse active:bg-[var(--ock-bg-inverse-active)] hover:bg-[var(--ock-bg-inverse-hover)]',
   primary:
     'cursor-pointer ock-bg-primary active:bg-[var(--ock-bg-primary-active)] hover:bg-[var(--ock-bg-primary-hover)]',
+  error: 'cursor-pointer ock-bg-error',
```
Curious why the error needs cursor-pointer? 

Why not just use the existing background.error?

rustam-cb replied:
Good point. Missed `background.error`. Updated now.

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,152 @@
+          'border-none bg-transparent',
+          'text-6xl leading-none outline-none',
+          'pointer-events-none absolute whitespace-nowrap opacity-0',
+        )}
```
with a long enough input, this span will trigger scrollbars on the page

rustam-cb replied:
Fixed

alessey commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +29,120 @@ export function FundButton({
+      case 'success':
+        return <SuccessSvg fill="#F9FAFB" />;
+      case 'error':
+        return <ErrorSvg fill="#F9FAFB" />;
```
should these use a theme variable?

rustam-cb replied:
No, we don't need the color to change based on the theme.

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,157 @@
+          onPopupClose={() => setSubmitButtonState('default')}
+        />
+      )}
+    </form>
```
I'm confused by this component design.  Its set up so you can pass in subcomponents, i.e. FundCardAmountInput, etc. but you can't really because of all the provider props they require.   How is this supposed to work?

rustam-cb replied:
This does not work. It did initially work the way I implemented it but I changed it to follow the other pattern in onchainkit and now it does not. I'll change it back. 
https://github.com/coinbase/onchainkit/pull/1718#discussion_r1892923451

rustam-cb replied:
Changed the customizable components to be passed as props

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,149 @@
+          PaymentMethodDropdownComponent={PaymentMethodDropdownComponent}
+          ButtonComponent={ButtonComponent}
+        />
+      </div>
```
if you pass in these components as children (and have a default children vs. default props), and rewrite the components to use the values from the providers instead of passing them from the provider as props, you give users the ability to customize layout and with className props more granularity over styling.  You do not really need the children loop, should be able to just pass them through.

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,89 @@
+  currencySign = '$',
+  paymentMethods = DEFAULT_PAYMENT_METHODS,
+  children,
+}: FundCardPropsReact) {
```
can we add a className prop to this and the sub components to allow consumers to customize?

rustam-cb replied:
Added

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,89 @@
+          <FundCardAmountInputTypeSwitch />
+
+          <FundCardPaymentMethodDropdown />
+        </>
```
this could be a prop default

rustam-cb replied:
Done

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,160 @@
+    [exchangeRate, setFundAmountFiat, setFundAmountCrypto, selectedInputType],
+  );
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: When value changes, we want to update the input width
```
nit: you could explicitly resize when value changes vs. listening for value updates

alessey commented on src/internal/svg/applePaySvg.tsx:
```diff
@@ -0,0 +1,23 @@
+      className={icon.foreground}
+    />
+  </svg>
+);
```
is this different then the appleSvg in the codebase?

rustam-cb replied:
Yes, those are different icons. apple icon vs applePay icon

alessey commented on src/fund/utils/formatDecimalInputValue.ts:
```diff
@@ -0,0 +1,21 @@
+    resultValue[1] !== '.'
+  ) {
+    resultValue = `${resultValue[0]}.${resultValue[1]}`;
+  }
```
this seems like slightly unintuitive behavior, might be better to replace 01 with 1 and just not allow 00

rustam-cb replied:
I feel like when user enters 01 they meant to have 0.1  and when they enter 00 they meant to enter 0.0 

I'll have this tested by the UX and will make any followup adjustments

rustam-cb commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,160 @@
+      // Set the input width based on the span width
+      if (inputRef.current) {
+        inputRef.current.style.width = `${width}px`;
+        inputRef.current.style.maxWidth = `${390 - currencyWidth}px`;
```
Updated

abcrane123 commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,88 @@
+          <div className={cn(background.inverse, 'overflow-y-auto p-2')}>
+            {paymentMethods.map((paymentMethod) => (
+              <FundCardPaymentMethodSelectRow
+                className={cn(background.inverse, 'px-4 py-2')}
```
I might've missed it but are you disabling debit card and apple pay if the amount is less than $5? the onramp team recently added this check so the onramp flow will error out if the amount is less than $5.

rustam-cb replied:
Addressed

abcrane123 commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,88 @@
+          )}
+        >
+          <div className={cn(background.inverse, 'overflow-y-auto p-2')}>
+            {paymentMethods.map((paymentMethod) => (
```
are you hiding apple pay on non safari / mobile devices?

rustam-cb replied:
Addressed

abcrane123 commented:
one last thing i forgot to ask about is the disabling of fields if < $5. You might want to check with design but for the Buy component we display a message that states minimum amount is $5 which i think is better than just hiding the row which could be confusing.

<img width="459" alt="Screenshot 2025-01-09 at 9 50 57 AM" src="https://github.com/user-attachments/assets/e6c6418e-e291-4e5f-aead-6b7d60cfb70a" />

abcrane123 commented:
i think the buy button is disabled here because i haven't connected wallet. you can probably do the same thing I did in the [Buy component ](https://github.com/coinbase/onchainkit/pull/1729/files#diff-25e2adaaccba43de700647b3d299088473ebba994e441797f2f4a8245c09415fR78)(replace buy button with connect wallet button if not connected) 

![Screenshot 2025-01-09 at 9 55 46 AM](https://github.com/user-attachments/assets/2b0f916c-ec7d-4df9-b068-ca0d7fee306e)

alessey commented:
can we add limits that correspond to the onramp limits?  i.e. 
![image](https://github.com/user-attachments/assets/c3a03f62-b877-4a7d-b56a-38cb2f7556b3)

alessey commented:
is this shadow under the dropdown correct?  
![image](https://github.com/user-attachments/assets/670e4441-9d45-47aa-a4b6-bdae0b6d629c)

alessey commented:
debit card says up to 500/week (but doesn't disable when i input > 500), but when it goes to onramp, it tells me up to 750.

alessey commented:
spacing between value and currency changes when you max out the input, may need to adjust the max width
![image](https://github.com/user-attachments/assets/e664413a-23a4-4587-a8b1-1942f9bba75f)

abcrane123 commented:
To follow the same pattern as our other components, can you use the `text` prop on `ConnectWallet` button to replace text with `Buy`

<img width="475" alt="Screenshot 2025-01-10 at 7 16 33 AM" src="https://github.com/user-attachments/assets/26c4da95-6bc2-4435-ba3a-389d8a467ea7" />

abcrane123 commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +30,121 @@ export function FundButton({
+  const fallbackFundingUrl = useGetFundingUrl();
+  const fundingUrlToRender = fundingUrl ?? fallbackFundingUrl;
   const isDisabled = disabled || !fundingUrlToRender;
+  const shouldShowConnectWallet = !fundingUrlToRender;
```
seems like this also will render it if `projectId` is undefined

abcrane123 commented:
i would also recommend posting the playground link in #proj-onchainkit-code-review so that the team / Tina / Mind can QA it for you before merging

alessey commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +30,121 @@ export function FundButton({
+        )}
+      </>
+    );
+  }, [buttonState, buttonIcon, buttonTextContent, hideText]);
```
nit: this might be more readable extracting one layer further so you can just have a single buttonState switch statement to return spinner/success/error/default

rustam-cb replied:
Will have that in a separate PR

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,94 @@
+        <FundCardAmountInputTypeSwitch />
+        <FundCardPaymentMethodDropdown />
+      </>
+    ),
```
nit: could make these an actual default prop

alessey commented on src/fund/hooks/useFundCardSetupOnrampEventListeners.ts:
```diff
@@ -0,0 +1,38 @@
+      unsubscribe();
+    };
+  }, []);
+};
```
nit: could make this more loosely coupled by calling a callback instead of tying directly to the fund context

rustam-cb replied:
There is already loosely coupled hook `setupOnrampEventListeners`. And this is just a wrapper to split out the code

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,156 @@
+    updateInputWidth();
+  }, [value, updateInputWidth]);
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: We only want to focus the input when the input type changes
```
would adding handleFocusInput to the deps and wrapping that in useCallback change anything?

rustam-cb replied:
Not sure what that would change but the current implementation works fine.

alessey commented on src/fund/components/FundCardProvider.tsx:
```diff
@@ -0,0 +1,118 @@
+  currencySign,
+  inputType,
+}: FundCardProviderReact) {
+  const [selectedAsset, setSelectedAsset] = useState<string>(asset);
```
does this need to be in state?  it only changes is solely based off the prop right?

alessey commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,140 @@
+  );
+
+  // If current selected method becomes disabled, switch to Coinbase
+  // biome-ignore lint/correctness/useExhaustiveDependencies: <explanation>
```
why is this ignore needed?

rustam-cb replied:
not needed

alessey commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,140 @@
+                testId={`ockFundCardPaymentMethodSelectRow__${paymentMethod.id}`}
+                paymentMethod={paymentMethod}
+                onClick={handlePaymentMethodSelect}
+                disabled={isPaymentMethodDisabled(paymentMethod)}
```
nit: since you are already calling isPaymentMethodDisabled you could conditionally apply the onClick using that as well

rustam-cb replied:
`FundCardPaymentMethodSelectRow` already does that

alessey commented on src/fund/components/FundCardPaymentMethodSelectorToggle.tsx:
```diff
@@ -0,0 +1,53 @@
+
+        <div className="relative flex items-center justify-center">
+          <div className="absolute top-0 left-0 h-4 w-4" />
+          {isOpen ? caretUpSvg : caretDownSvg}
```
nit: would be a little nicer to use the same svg and rotate it with an animation

alessey commented on src/fund/components/FundCardProvider.tsx:
```diff
@@ -0,0 +1,118 @@
+    );
+  }, 1000);
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: One time effect
```
if you only call fetchExchangeRate once, why is it debounced?

rustam-cb replied:
It was for future implementation, but removed for now.

alessey commented on src/fund/components/FundCardProvider.tsx:
```diff
@@ -0,0 +1,118 @@
+    setExchangeRateLoading,
+    submitButtonState,
+    setSubmitButtonState,
+    paymentMethods: paymentMethods || DEFAULT_PAYMENT_METHODS,
```
how would this be used?  A user could restrict the payment methods based on DEFAULT_PAYMENT_METHODS?  this code seems pretty tied to the defaults, wouldn't other functionality like the payment select break if there was no Coinbase method?

alessey commented on src/fund/utils/truncateDecimalPlaces.ts:
```diff
@@ -0,0 +1,12 @@
+ * Limit the value to N decimal places
+ */
+export const truncateDecimalPlaces = (value: string, decimalPlaces: number) => {
+  const decimalIndex = value.indexOf('.');
```
nit: could easily make this more flexible by allowing value to be string | number

alessey commented on src/internal/svg/successSvg.tsx:
```diff
@@ -11,7 +11,7 @@ export const successSvg = (
     <path
       d="M8 0C3.58 0 0 3.58 0 8C0 12.42 3.58 16 8 16C12.42 16 16 12.42 16 8C16 3.58 12.42 0 8 0ZM6.72667 11.5333L3.73333 8.54L4.67333 7.6L6.72667 9.65333L11.44 4.94L12.38 5.88L6.72667 11.5333Z"
-      fill="#65A30D"
+      fill={fill}
```
nit: do the success and error icons really need to be different colors between ock components?

rustam-cb replied:
These icons are used in other components with the color that is different than what I need.

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,94 @@
+  assetSymbol,
+  buttonText = 'Buy',
+  headerText,
+  currencySign = '$',
```
do we support any other currencies?

rustam-cb replied:
We do. I'll have supporting different currencies in a followup separate PR

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,94 @@
+        state={submitButtonState}
+        onClick={() => setSubmitButtonState('loading')}
+        onPopupClose={() => setSubmitButtonState('default')}
+      />
```
shouldn't the FundButton be set up the same as the other children?

alessey commented on src/fund/components/FundCard.tsx:
```diff
@@ -0,0 +1,94 @@
+          componentTheme,
+          background.default,
+          color.foreground,
+          'flex w-[440px] flex-col p-6',
```
isn't 440px to large for most mobile devices?

alessey commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,156 @@
+          onChange={handleChange}
+          ref={inputRef}
+          inputMode="decimal"
+          minLength={1}
```
minLength is not a valid attribute for type="number"

alessey commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,140 @@
+                onClick={handlePaymentMethodSelect}
+                disabled={isPaymentMethodDisabled(paymentMethod)}
+                disabledReason={
+                  isPaymentMethodDisabled(paymentMethod)
```
extract isPaymentMethodDisabled to a const in this map

alessey commented on src/fund/components/FundCardPaymentMethodDropdown.tsx:
```diff
@@ -0,0 +1,140 @@
+          <div className={cn(background.inverse, 'overflow-y-auto p-2')}>
+            {filteredPaymentMethods.map((paymentMethod) => (
+              <FundCardPaymentMethodSelectRow
+                className={cn(background.inverse, 'px-4 py-2')}
```
any reason not to just apply these styles in the FundCardPaymentMethodSelectRow so they are all in once place?  it appears you are overriding styles set in that component even though its only used here

rustam-cb replied:
Moved.

alessey commented:
this component should implement a lifecycle status to provide onStatus, onSuccess, onError callbacks: https://onchainkit.xyz/guides/lifecycle-status

rustam-cb commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+    formatCrypto,
+  ]);
+
+  if (exchangeRateLoading || !exchangeRate) {
```
This is going to be handled in a follow up PRs

alessey commented on src/fund/components/FundCardCurrencyLabel.tsx:
```diff
@@ -0,0 +1,23 @@
+        'flex items-center justify-center bg-transparent',
+        'text-6xl leading-none outline-none',
+      )}
+      data-testid="currencySpan"
```
nit: ockCurrencySpan

rustam-cb replied:
updated

alessey commented on src/fund/components/FundCardHeader.tsx:
```diff
@@ -0,0 +1,13 @@
+  const { headerText } = useFundContext();
+
+  return (
+    <div className={cn(text.headline, className)} data-testid="fundCardHeader">
```
nit: testIds are generally prefixed with ock, i.e. ockFundCardHeader

rustam-cb replied:
updated

abcrane123 commented on src/fund/components/FundCardAmountInput.tsx:
```diff
@@ -0,0 +1,148 @@
+  className,
+}: FundCardAmountInputPropsReact) => {
+  // TODO: Get currency label from country
+  const currencyLabel = 'USD';
```
is this TODO a follow up ?

rustam-cb replied:
Yes, its going to be in the follow up PRs.

abcrane123 commented on src/internal/svg/creditCardSvg.tsx:
```diff
@@ -0,0 +1,23 @@
+    fill="none"
+    xmlns="http://www.w3.org/2000/svg"
+  >
+    <path
```
title here as well

abcrane123 replied:
is this the same as the buy component svg ? if so we already have one `cardSvg`. same for apple pay

abcrane123 commented on src/internal/svg/addSvg.tsx:
```diff
@@ -13,7 +13,7 @@ export const addSvg = (
     <path
       d="M7.125 0H5.875V5.37501L0.5 5.37501L0.5 6.62501H5.875V12H7.125V6.62501H12.5V5.37501H7.125V0Z"
-      className={icon.inverse}
+      className={className}
```
can you add a title here plz

brendan-defi commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +31,122 @@ export function FundButton({
   );
 
+  const buttonColorClass = useMemo(() => {
+    switch (buttonState) {
+      case 'error':
+        return background.error;
+      case 'loading':
+      case 'success':
+        return pressable.primary;
+      default:
+        return pressable.primary;
+    }
```
are there cases where you can envision returning anything other than background.error or pressable.primary?

unless you envision a whole bunch of different styles, it would match our style better to do something like:
``` 
if (error) {
  return background.error;
}
return pressable.primary;
```

brendan-defi commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +31,122 @@ export function FundButton({
-    'px-4 py-3 inline-flex items-center justify-center space-x-2 disabled',
+    buttonColorClass,
+    'px-4 py-3 inline-flex items-center justify-center space-x-2',
     isDisabled && pressable.disabled,
```
the idiomatic way to write this using the `cn` utility is with object syntax:
{
  pressable.disabled: isDisabled,
}

brendan-defi commented on src/fund/components/FundButton.tsx:
```diff
@@ -19,49 +31,122 @@ export function FundButton({
+
   const classNames = cn(
     componentTheme,
-    pressable.primary,
-    'px-4 py-3 inline-flex items-center justify-center space-x-2 disabled',
+    buttonColorClass,
+    'px-4 py-3 inline-flex items-center justify-center space-x-2',
     isDisabled && pressable.disabled,
     text.headline,
     border.radius,
     color.inverse,
     className,
   );
```
the idiomatic way to write conditional values using the cn utility is with object syntax:
{
pressable.disabled: isDisabled,
}

brendan-defi commented on src/fund/components/FundCardAmountInputTypeSwitch.tsx:
```diff
@@ -0,0 +1,99 @@
+    setSelectedInputType(selectedInputType === 'fiat' ? 'crypto' : 'fiat');
+  };
+
+  const formatUSD = useCallback((amount: string) => {
+    const roundedAmount = Number(getRoundedAmount(amount || '0', 2));
+    return `$${roundedAmount}`;
+  }, []);
+
+  const formatCrypto = useCallback(
+    (amount: string) => {
+      return `${truncateDecimalPlaces(amount || '0', 8)} ${asset}`;
+    },
+    [asset],
+  );
```
mentioned above a method that I think is probably better for crypto. here's a fiat version:
```
new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  })
```

rustam-cb replied:
I'll keep the currently implemented solution for this iteration

rustam-cb replied:
Will have this updated in follow up PRs

brendan-defi commented on src/fund/components/FundCardSubmitButton.tsx:
```diff
@@ -0,0 +1,28 @@
+      className="w-full"
+      fundingUrl={fundingUrl}
+      state={submitButtonState}
+      onClick={() => setSubmitButtonState('loading')}
+      onPopupClose={() => setSubmitButtonState('default')}
```
even for simple functions, i believe we prefer callbacks here

rustam-cb replied:
using `useCallback` should not be preferred in this case as its more cleaner this way and using useCallback in this case would not provide any benefits.

brendan-defi commented on src/internal/svg/addSvg.tsx:
```diff
@@ -11,9 +11,10 @@ export const addSvg = (
     fill="none"
     xmlns="http://www.w3.org/2000/svg"
   >
+    <title>Add SVG</title>
```
this is what is read when accessibility settings are on, so we prefer to make the title "nicer". Recommend keeping this as "add"

brendan-defi commented on src/internal/svg/appleSvg.tsx:
```diff
@@ -1,29 +1,24 @@
-    height="24"
   >
-    <title>Apple Pay Onramp</title>
+    <title>Apple SVG</title>
```
same comment about titles. prefer to remove SVG and make as intuitive as possible when it's read out loud

brendan-defi commented on src/internal/svg/creditCardSvg.tsx:
```diff
@@ -0,0 +1,24 @@
+    fill="none"
+    xmlns="http://www.w3.org/2000/svg"
+  >
+    <title>Credit Card SVG</title>
```
same comment about titles

brendan-defi commented on src/internal/svg/errorSvg.tsx:
```diff
@@ -11,7 +11,7 @@ export const errorSvg = (
     <path
       d="M8 16C12.4183 16 16 12.4183 16 8C16 3.58171 12.4183 0 8 0C3.58172 0 0 3.58171 0 8C0 12.4183 3.58172 16 8 16ZM11.7576 5.0909L8.84853 8L11.7576 10.9091L10.9091 11.7576L8 8.84851L5.09093 11.7576L4.2424 10.9091L7.15147 8L4.2424 5.0909L5.09093 4.24239L8 7.15145L10.9091 4.24239L11.7576 5.0909Z"
-      fill="#E11D48"
+      fill={fill}
```
is it expected that this would take values that are not in our style system? 

would prefer to do this via `background.error` or `background.warning` by importing background from `src/styles`. 

additional benefit of not needing to make this a function

rustam-cb replied:
This icon is being used by other component with the specific color `#E11D48`. So I'm just making this to work for me at this point. Somebody else did not like that approach  because it was not supposed to change colors with theme.

brendan-defi commented on src/styles/theme.ts:
```diff
@@ -1,5 +1,5 @@
-import { clsx } from 'clsx';
 import type { ClassValue } from 'clsx';
+import { clsx } from 'clsx';
```
nit: can combine these into one import line

abcrane123 commented:
could you add the lifecycle callbacks to the demo in playground so we can check those are working as expected? i.e onStatus, onSuccess, onError

alessey commented:
inputs kinda wonky

https://github.com/user-attachments/assets/3e430f7d-f6b7-4677-9a37-0391012af603

alessey commented:
possible to follow the pattern defined in our docs for lifecycle status?
https://onchainkit.xyz/guides/lifecycle-status

cpcramer commented:
Not a blocker but currently the scroll wheel updates the input price - this feels odd
